File: BinUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LVLTool
{
    public static class BinUtils
    {
        private static bool Find(byte[] hexNumber, long location, byte[] data)
        {
            int index = 0;
            while (index < hexNumber.Length && (int)hexNumber[index] == (int)data[location + (long)index])
                ++index;
            return index == hexNumber.Length;
        }

        private static bool Find(byte[] hexNumber, long location, List<byte> data)
        {
            int index = 0;
            while (index < hexNumber.Length && (int)hexNumber[index] == (int)data[(int)(location + index)])
                ++index;
            return index == hexNumber.Length;
        }

        /// <summary>
        /// Searched through 'data' for the givenBytes.
        /// </summary>
        /// <param name="startLocation">Where to start in 'data'</param>
        /// <param name="givenBytes">the stuff to look for</param>
        /// <param name="data">the big data to look through</param>
        /// <returns>location of the givenBytes, -1L if not found.</returns>
        public static long GetLocationOfGivenBytes(long startLocation, byte[] givenBytes, byte[] data, long maxDistance)
        {
            long retVal = -1L;
            long num = (long)(data.Length - givenBytes.Length);
            for (long location = startLocation; location < num; ++location)
            {
                if (Find(givenBytes, location, data))
                {
                    retVal = location;
                    break;
                }
                else if (maxDistance < location - startLocation)
                    break;
            }
            return retVal;
        }

        public static long GetLocationOfGivenBytes(long startLocation, byte[] givenBytes, List<byte> data, long maxDistance)
        {
            long retVal = -1L;
            long num = (long)(data.Count - givenBytes.Length);
            for (long location = startLocation; location < num; ++location)
            {
                if (Find(givenBytes, location, data))
                {
                    retVal = location;
                    break;
                }
                else if (maxDistance < location - startLocation)
                    break;
            }
            return retVal;
        }

        public static string GetByteString(long location, byte[] data, int numBytes)
        {
            char c = '.';
            if (location + numBytes > data.Length)
                return "";
            StringBuilder b = new StringBuilder(numBytes);
            if (location < 0) location = 0;
            
            for (long l = location; l < location + numBytes; l++)
            {
                if (data[l] != 0)
                {
                    c = (char)data[l];
                    b.Append(c);
                }
            }
            return b.ToString();
        }
        // The number/length is stored in reverse least significant --> most significant
        // This function gets the number at the specified location and gives it to you good ;
        public static UInt16 Get2ByteNumberAtLocation(int loc, List<byte> data)
        {
            byte b0, b1;
            b0 = data[loc];
            b1 = data[loc + 1];

            UInt16 retVal = (UInt16)(b0 + (b1 << 8));
            return retVal;
        }

        // The number/length is stored in reverse least significant --> most significant
        // This function gets the number at the specified location and gives it to you good ;
        public static uint GetNumberAtLocation(int loc, List<byte> data)
        {
            byte b0, b1, b2, b3;
            b0 = data[loc];
            b1 = data[loc + 1];
            b2 = data[loc + 2];
            b3 = data[loc + 3];

            uint retVal = (uint)(b0 + (b1 << 8) + (b2 << 16) + (b3 << 24));
            return retVal;
        }

        // The number/length is stored in reverse least significant --> most significant
        // This function gets the number at the specified location and gives it to you good ;
        public static uint GetNumberAtLocation(long loc, byte[] data)
        {
            byte b0, b1, b2, b3;
            b0 = data[loc];
            b1 = data[loc + 1];
            b2 = data[loc + 2];
            b3 = data[loc + 3];

            uint retVal = (uint)( b0 + (b1 << 8) + (b2 << 16) + (b3 << 24));
            return retVal;
        } 

        // The number/length is stored in reverse least significant --> most significant
        // This function sets the number at the specified location 
        public static void WriteNumberAtLocation(long loc, UInt32 num, byte[] data)
        {
            byte[] encodedNumber = EncodeNumber( (int) num);

            data[loc] = encodedNumber[0];
            data[loc + 1] = encodedNumber[1];
            data[loc + 2] = encodedNumber[2];
            data[loc + 3] = encodedNumber[3];
        }

        // The number/length is stored in reverse least significant --> most significant
        // This function sets the number at the specified location
        public static void WriteNumberAtLocation(int loc, UInt32 num, List<byte> data)
        {
            byte[] encodedNumber = EncodeNumber((int)num);

            data[loc] = encodedNumber[0];
            data[loc + 1] = encodedNumber[1];
            data[loc + 2] = encodedNumber[2];
            data[loc + 3] = encodedNumber[3];
        }

        // The number/length is stored in reverse least significant --> most significant
        // This function sets the number at the specified location
        public static void Write2ByteNumberAtLocation(int loc, UInt16 num, List<byte> data)
        {
            byte[] encodedNumber = EncodeNumber((int)num);

            data[loc] = encodedNumber[0];
            data[loc + 1] = encodedNumber[1];
        }

        public static byte[] EncodeNumber(int num)
        {
            byte[] retVal = new byte[4];
            retVal[0] = (byte)(num & 0x000000FF); // last byte
            retVal[1] = (byte)((num & 0x0000FF00) >> 8);
            retVal[2] = (byte)((num & 0x00FF0000) >> 16);
            retVal[3] = (byte)((num & 0xFF000000) >> 24); // first byte
            return retVal;
        }

        public static bool BinCompare(byte[] littleData, byte[] bigData, long location)
        {
            for (long i = 0; i < littleData.Length; i++)
                if (littleData[i] != bigData[i + location])
                    return false;
            return true;
        }
    }
}

File: dart_translation/lib/bin_utils.dart
/// Binary utilities for LVL/UCFB data.
/// Translated from C# CoreCode/BinUtils.cs for Linux/CLI use.
/// Byte data uses [Uint8List]; [List<int>] parameters accept [Uint8List] (implements [List<int>]).

import 'dart:typed_data';

/// Searches through [data] for [givenBytes].
/// Returns location of givenBytes, -1 if not found.
int getLocationOfGivenBytes(
    int startLocation, List<int> givenBytes, List<int> data, int maxDistance) {
  int retVal = -1;
  int num = data.length - givenBytes.length;
  for (int location = startLocation; location < num; location++) {
    if (_find(givenBytes, location, data)) {
      retVal = location;
      break;
    } else if (maxDistance < location - startLocation) {
      break;
    }
  }
  return retVal;
}

bool _find(List<int> hexNumber, int location, List<int> data) {
  int index = 0;
  while (index < hexNumber.length &&
      hexNumber[index] == data[location + index]) {
    index++;
  }
  return index == hexNumber.length;
}

/// Same for Uint8List (byte array).
int getLocationOfGivenBytesFromBytes(
    int startLocation, List<int> givenBytes, List<int> data, int maxDistance) {
  return getLocationOfGivenBytes(startLocation, givenBytes, data, maxDistance);
}

String getByteString(int location, List<int> data, int numBytes) {
  if (location + numBytes > data.length) return '';
  if (location < 0) location = 0;
  final buffer = StringBuffer();
  for (int l = location; l < location + numBytes; l++) {
    if (data[l] != 0) {
      buffer.write(String.fromCharCode(data[l]));
    }
  }
  return buffer.toString();
}

/// Number/length is stored little-endian (LSB first). Result as 16-bit unsigned.
int get2ByteNumberAtLocation(int loc, List<int> data) {
  int b0 = data[loc] & 0xff;
  int b1 = data[loc + 1] & 0xff;
  return (b0 + (b1 << 8)) & 0xFFFF;
}

/// 4-byte number at location (little-endian). Result as 32-bit unsigned.
int getNumberAtLocation(int loc, List<int> data) {
  int b0 = data[loc] & 0xff;
  int b1 = data[loc + 1] & 0xff;
  int b2 = data[loc + 2] & 0xff;
  int b3 = data[loc + 3] & 0xff;
  return (b0 + (b1 << 8) + (b2 << 16) + (b3 << 24)) & 0xFFFFFFFF;
}

void writeNumberAtLocation(int loc, int num, List<int> data) {
  Uint8List encoded = encodeNumber(num);
  data[loc] = encoded[0];
  data[loc + 1] = encoded[1];
  data[loc + 2] = encoded[2];
  data[loc + 3] = encoded[3];
}

void write2ByteNumberAtLocation(int loc, int num, List<int> data) {
  Uint8List encoded = encodeNumber(num);
  data[loc] = encoded[0];
  data[loc + 1] = encoded[1];
}

/// Encodes int as 4 little-endian bytes.
Uint8List encodeNumber(int num) {
  return Uint8List(4)
    ..[0] = num & 0x000000FF
    ..[1] = (num & 0x0000FF00) >> 8
    ..[2] = (num & 0x00FF0000) >> 16
    ..[3] = (num & 0xFF000000) >> 24;
}

bool binCompare(List<int> littleData, List<int> bigData, int location) {
  for (int i = 0; i < littleData.length; i++) {
    if (littleData[i] != bigData[i + location]) return false;
  }
  return true;
}
